docs/development/test-guidelines.adoc

```diff
@@ -0,0 +1,209 @@
+<1> Notice the package name is just `bar` so we have access to all symbols (exported and unexported ones).
+<2> Test the unexported symbol `toUpper`.
+<3> Marking the test for the `toUpper` function test as: "executable in parallel".
+
+The whitebox test <<lib-whitebox-test-dot-go-source>> uses the `bar` package name
+and thereby gets access to all the internal of a package. There is also no need
+to import the `foo/bar` package. We can just use the `toUpper()` function.
+
+In order to confirm that our code actually works, let's jump to a shell and run these commands:
+
+[source,bash]
+.Run tests and collect coverage information
+----
+$ go test -v -coverprofile cov.out foo/bar # <1>
+=== RUN   TestToUpper
+--- PASS: TestToUpper (0.00s)
+=== RUN   TestSay
+=== RUN   TestShout
+--- PASS: TestSay (0.00s)
+--- PASS: TestShout (0.00s)
+PASS
+coverage: 100.0% of statements
+ok  	foo/bar	0.002s
+
+
+$ go tool cover -func=cov.out # <2>
+foo/bar/lib.go:9:	Say		100.0%
+foo/bar/lib.go:15:	Shout		100.0%
+foo/bar/lib.go:20:	toUpper		100.0%
+total:			(statements)	100.0%
+----
+<1> Runs all the tests in the `foo/bar` package and collects coverage information in the `cov.out` file.
+<2> Prints coverage information per function in the package.
+
+[[why-no-build-tags]]
+== Why don't we use build tags?
+
+What we want is to compile all code in our codebase to find errors quickly.
+
+At runtime we may want to skip some tests for speed or missing dependency reasons. 
+
+There are people on the internet that propose the use of build tags to distinguish
+unit and integration tests. The way this works is as follows:
+
+ 1. You write a comment in the form of `// +build unit` or `// +build integration`
+    as the first line inside of a `*.go` file.
+ 2. As a consequence such a file will only be compiled by `go build` or `go test`
+    if a special tag (e.g. `unit` or `integration`) is specified on the command
+    line using the `-tags` parameter (see `go build --help`).
+
+This has dramatic side effects:
+
+Suppose all that you wanted was to distinguish *at runtime* if a test shall be run or not (skipped).
+With build tags though, your code will *not even be compiled* and *compile errors cannot be found*
+if a tag is missing.
+
+That is why we try to avoid build tags. We don't say that build tags are completely useless.
+In order to compile platform specific code, they come in handy for example.
+
+[[why-no-short-parameter]]
+== Why don't we use the -short parameter?
+
+The `go test` command offers a switch ( `-short` ) to skip long running tests,
+that have been marked as such.
+
+The downside of build tags (see <<why-no-build-tags>>) is eliminated because
+the decision whether a test is run, is made at runtime.
+
+The only reason why we still don't use the `-short` parameter is that we want to
+have the ability to define more test-skipping criteria than just the time
+consumed by a test.
+That is why we have the `resource.require(t, resource.Database)` syntax to
```

Comment:
I'd add `require` to the `test` package where it should belong. But that can be done afterwards.